Applications/Spire/Source/KeyBindings/OrderTaskArgumentsMatch.cpp

```diff
@@ -0,0 +1,177 @@
+#include "Nexus/Definitions/DefaultDestinationDatabase.hpp"
+#include "Spire/Ui/CustomQtVariants.hpp"
+
+using namespace boost;
+using namespace Nexus;
+using namespace Spire;
+
+namespace {
+  bool matches_by_word(const QString& text, const QString& query) {
+    auto words = text.splitRef(' ', Qt::SkipEmptyParts);
+    for(auto& word : words) {
+      if(text.midRef(
+          text.indexOf(word)).startsWith(query, Qt::CaseInsensitive)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+bool Spire::matches(const QString& name, const QString& query) {
+  for(auto i = 0; i < name.length(); ++i) {
+    if(name.right(name.length() - i).startsWith(query, Qt::CaseInsensitive)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool Spire::matches(const CountryCode& country, const QString& query) {
+  auto& entry = GetDefaultCountryDatabase().FromCode(country);
+  return entry.m_code != CountryCode::NONE &&
+    (QString::fromStdString(entry.m_twoLetterCode.GetData()).startsWith(
+        query, Qt::CaseInsensitive) ||
+      QString::fromStdString(entry.m_threeLetterCode.GetData()).startsWith(
+        query, Qt::CaseInsensitive) ||
+      QString::fromStdString(entry.m_name).startsWith(
+        query, Qt::CaseInsensitive));
+}
+
+bool Spire::matches(const MarketCode& market, const QString& query) {
+  auto& entry = GetDefaultMarketDatabase().FromCode(market);
+  return !entry.m_code.IsEmpty() &&
+    (QString::fromStdString(entry.m_code.GetData()).startsWith(
+        query, Qt::CaseInsensitive) ||
+      QString::fromStdString(entry.m_displayName).startsWith(
+        query, Qt::CaseInsensitive) ||
+      matches_by_word(QString::fromStdString(entry.m_description), query));
+}
+
+bool Spire::matches(const Security& security, const QString& query) {
+  return !security.GetSymbol().empty() &&
+    (to_text(security).startsWith(query, Qt::CaseInsensitive) ||
+      matches(security.GetMarket(), query));
+}
+
+bool Spire::matches(const Region& region, const QString& query) {
+  if(!region.GetName().empty() && QString::fromStdString(
+      region.GetName()).startsWith(query, Qt::CaseInsensitive)) {
+    return true;
+  }
+  for(auto& country : region.GetCountries()) {
+    if(matches(country, query)) {
+      return true;
+    }
+  }
+  for(auto& market : region.GetMarkets()) {
+    if(matches(market, query)) {
+      return true;
+    }
+  }
+  for(auto& security : region.GetSecurities()) {
+    if(matches(security, query)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool Spire::matches(const Destination& destination, const QString& query) {
+  auto& entry = GetDefaultDestinationDatabase().FromId(destination);
+  return !entry.m_id.empty() &&
+    (QString::fromStdString(entry.m_id).startsWith(
+        query, Qt::CaseInsensitive) ||
+      matches_by_word(QString::fromStdString(entry.m_description), query));
+}
+
+bool Spire::matches(const OrderType& order_type, const QString& query) {
```

Comment:
Pay attention to whether an argument is passed by reference or passed by value:

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const